### ISSUE TITLE
C++ and python example

### ISSUE THREAD

**Issue by din14970** [berkels/match-series#4]
Here is an example of a C++/CUDA code with python bindings:
https://github.com/prism-em/prismatic
Everything is in the same repository.
I have no idea how it is done.

**Comment by berkels** [comment]
Thanks for the info! I did a quick search and noticed that the structure looks similar to https://github.com/pybind/cmake_example. So perhaps this is based on pybind11. Definitely something to look into.

**Comment by berkels** [comment]
Just a quick update: pybind11 seems to be a reasonable way to create bindings for a handful of C++ functions from the match-series code that we may want to call directly from Python. Including the C++ headers in the install target will help with this, which is why I added this in 1.0.3.

**Comment by din14970** [comment]
This would be very helpful, writing a wrapper around just the monolithic binary is a bit clunky

**Comment by berkels** [comment]
As a proof of concept, I added a python binding for one of the registration functions from match-series using pybind11:
https://github.com/berkels/match-series-pybind11-example, I used https://github.com/pybind/cmake_example as base for this.

**Comment by din14970** [comment]
If I knew C++ I would give it a go, but looking through your codebase I don't have much hope this is a wise investment of my time. The function one would care about is still the one invoked by the `matchSeries` command, so I suppose that is just the `main` function. It would be most convenient if the C++ stuff came bundled with the python wrapper as opposed to requiring a separate install, but for now it is acceptable. 

I have spent a few days improving the python wrapper; once I've written some unit tests I will let you know about the new release and you can check it out. All the annoying stuff is now abstracted away and in my opinion everything is quite pythonic. The interactive workflow is simply:

```python
from pymatchseries import MatchSeries

calculation = MatchSeries(data)  # data can be a numpy array, dask array or hyperspy Signal2D
calculation.configuration["parameter"]=value  # most configuration is auto-filled but values can be changed in this way
calculation.run()  # a standard folder structure is created and the calculation is started with the output piped to stdout
deformed_images = calculation.get_deformed_images()
deformed_spectra = calculation.apply_deformation_to_spectra(spectrum_map)
```

Something like this would allow for easy parameter studies. Under the hood I am now using [dask](https://docs.dask.org/en/latest/) to deal with the mapping of the deformations to the datasets ideally as efficiently and parallelized as possible. Dask also allows me to define all operations lazily before computing them, ensuring that I am never forced to load huge datasets into memory. 

The truly ideal scenario would be if the match-series code were also implemented in python. The past few months I've written a few things using numba and the results are really incredible. Numba also works really well with Dask, allows you to autoparallelize and even compile for the GPU. I've also used a [TEM image simulation code](https://abtem.readthedocs.io/en/latest/install.html) written entirely in python using numba and cupy on our machine learning cluster to do essentially high-throughput STEM simulation, and it's basically as fast and less error prone than [a comparable C++ code](https://prism-em.com/). Do you think it would be difficult (i.e. a significant time investment) to implement the algorithm entirely in python? How much potential is there for speed-up and parallelization? Looking through the repo there seems to be a lot in here, but I'm wondering how much is boilerplate that could be done with a few lines of python.

**Comment by berkels** [comment]
> If I knew C++ I would give it a go, but looking through your codebase I don't have much hope this is a wise investment of my time.

I agree. The C++ codebase is too complex und not sufficiently documented for this.

> The function one would care about is still the one invoked by the matchSeries command, so I suppose that is just the main function. It would be most convenient if the C++ stuff came bundled with the python wrapper as opposed to requiring a separate install, but for now it is acceptable.

For just exporting `main`, I'd say pybind11 is overkill. I was thinking in the direction of exporting some inner building blocks so that the outer logic can be done in Python. For instance, exporting the function that takes an image pair, an initial guess for the non-rigid deformation and returns the non-rigid deformation that aligns the image pair.

> I have spent a few days improving the python wrapper; once I've written some unit tests I will let you know about the new release and you can check it out. All the annoying stuff is now abstracted away and in my opinion everything is quite pythonic. 

Sounds great.

> The truly ideal scenario would be if the match-series code were also implemented in python.

In the notation of https://arxiv.org/pdf/1403.6774.pdf, implementing Algorithms 1 and 2 in Python is rather simple (except for one thing). I actually have Python variants of those algorithms for rigid registration and a simplified discretization. What would need some time investment is to implement the functions needed to evaluate equation (7), i.e. the discretized objective function, its gradient and the inverse of A. Should be surely doable efficiently with numba, but definitely needs more than an afternoon of work ;-).

> How much potential is there for speed-up and parallelization?

On a single core, I'd say the C++ code is rather efficient and I don't expect you would gain much doing this with numba. There is very large potential for speed-up by parallelization though. In principle, this could scale almost linearly with the number of cores and is also GPU friendly since you can parallelize over the pixels. The C++ code doesn't make use of this, since it's written in an abstract form that allows for much more general discretizations, e.g. arbitrary triangle meshes. Since we only need cartesian grids, this simplifies things greatly. If you are interested, we can have a video meeting to discuss this in more detail.

**Comment by din14970** [comment]
I've created a new version: https://pypi.org/project/pyMatchSeries/  
and updated the example: https://github.com/din14970/pyMatchSeries/blob/master/examples/example.ipynb
I think it works a lot better now, application to spectrum maps is scaleable, and it is more tightly integrated with hyperspy, which is commonly used in the electron microscopy community. It's better maintainable and now has automatically running unit tests via github actions, so if you break the binary for some platform I will know about it ;). 

> For instance, exporting the function that takes an image pair, an initial guess for the non-rigid deformation and returns the non-rigid deformation that aligns the image pair.

Yes you are right, only I suspect bundling and building will become quite complicated.

> What would need some time investment is to implement the functions needed to evaluate equation (7), i.e. the discretized objective function, its gradient and the inverse of A. Should be surely doable efficiently with numba, but definitely needs more than an afternoon of work ;-).

Is this a custom algorithm or can this be taken from somewhere?

> If you are interested, we can have a video meeting to discuss this in more detail.

I'm open to a discussion anytime, but I'm not sure whether I will have time to program it myself - GPU programming is a bit of a craft of its own. I think it would be great to have this be useable as one could then run it on large datasets. I'm mainly concerned with STEM-EDX datasets which have lots and lots of images, and there are quite a few papers out there that show the spectral results can be significantly improved, but they don't usually share very useable code. If we could run this fast and parallelized on our GPU cluster one could really do extensive parameter studies.

**Comment by berkels** [comment]
>  It's better maintainable and now has automatically running unit tests via github actions, so if you break the binary for some platform I will know about it ;).

Sounds great!

> Yes you are right, only I suspect bundling and building will become quite complicated.

Yeah, I think figuring out the build process may get complicated. I would assume that we only need to do this once though. Whether or not its worth the time investment is difficult to say. If we mange to get a Python implementation in the not too distant future, there is surely no need to invest time in Python bindings for the C++ code.

> Is this a custom algorithm or can this be taken from somewhere?

The ingredients are pretty much standard, but I'm not aware of a Python implementation that has the combination we need.

> I'm open to a discussion anytime, but I'm not sure whether I will have time to program it myself - GPU programming is a bit of a craft of its own. 

A basic numpy/numba CPU implementation would already be great, no need for GPU support. The main problem I have with the C++ code is that it's very difficult to extend and very difficult to even understand, if you haven't seen quocmesh before. A Python implementation with speed comparable to the existing (more or less single core) C++ implementation would solve that. If you are available next week, it would be great if we can discuss this in more detail.

**Comment by din14970** [comment]
Sure next week is fine for me more or less anytime. Let me know what fits you best.
